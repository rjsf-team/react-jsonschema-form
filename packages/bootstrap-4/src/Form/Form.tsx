import { withTheme, FormProps } from "@rjsf/core";

import Theme from "../Theme";
import { StatelessComponent } from "react";

const Form:
  | React.ComponentClass<FormProps<any>>
  | StatelessComponent<FormProps<any>> = withTheme(Theme);

export default Form;
