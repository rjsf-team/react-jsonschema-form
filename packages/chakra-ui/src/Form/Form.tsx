import { withTheme, FormProps } from "@rjsf/core";

import Theme from "../Theme";
import { ComponentClass, FunctionComponent } from "react";

const Form:
  | ComponentClass<FormProps<any>>
  | FunctionComponent<FormProps<any>> = withTheme(Theme);

export default Form;
