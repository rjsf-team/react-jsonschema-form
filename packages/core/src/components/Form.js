import React, { Component } from "react";
import PropTypes from "prop-types";
import _pick from "lodash/pick";
import _get from "lodash/get";
import _isEmpty from "lodash/isEmpty";

import { default as DefaultErrorList } from "./ErrorList";
import {
  getDefaultFormState,
  retrieveSchema,
  shouldRender,
  toIdSchema,
  getDefaultRegistry,
  deepEquals,
  toPathSchema,
  isObject,
} from "../utils";
import validateFormData, { toErrorList } from "../validate";
import { mergeObjects } from "../utils";

export default class Form extends Component {
  static defaultProps = {
    uiSchema: {},
    noValidate: false,
    liveValidate: false,
    disabled: false,
    readonly: false,
    noHtml5Validate: false,
    ErrorList: DefaultErrorList,
    omitExtraData: false,
  };

  constructor(props) {
    super(props);
    this.state = this.getStateFromProps(props, props.formData);
    if (
      this.props.onChange &&
      !deepEquals(this.state.formData, this.props.formData)
    ) {
      this.props.onChange(this.state);
    }
    this.formElement = null;
  }

  UNSAFE_componentWillReceiveProps(nextProps) {
    const nextState = this.getStateFromProps(nextProps, nextProps.formData);
    if (
      !deepEquals(nextState.formData, nextProps.formData) &&
      !deepEquals(nextState.formData, this.state.formData) &&
      this.props.onChange
    ) {
      this.props.onChange(nextState);
    }
    this.setState(nextState);
  }

  getStateFromProps(props, inputFormData) {
    const state = this.state || {};
    const schema = "schema" in props ? props.schema : this.props.schema;
    const uiSchema = "uiSchema" in props ? props.uiSchema : this.props.uiSchema;
    const edit = typeof inputFormData !== "undefined";
    const liveValidate =
      "liveValidate" in props ? props.liveValidate : this.props.liveValidate;
    const mustValidate = edit && !props.noValidate && liveValidate;
    const rootSchema = schema;
    const formData = getDefaultFormState(schema, inputFormData, rootSchema);
    const retrievedSchema = retrieveSchema(schema, rootSchema, formData);
    const customFormats = props.customFormats;
    const additionalMetaSchemas = props.additionalMetaSchemas;

    const getCurrentErrors = () => {
      if (props.noValidate) {
        return { errors: [], errorSchema: {} };
      } else if (!props.liveValidate) {
        return {
          errors: state.schemaValidationErrors || [],
          errorSchema: state.schemaValidationErrorSchema || {},
        };
      }
      return {
        errors: state.errors || [],
        errorSchema: state.errorSchema || {},
      };
    };

    let errors,
      errorSchema,
      schemaValidationErrors,
      schemaValidationErrorSchema;
    if (mustValidate) {
      const schemaValidation = this.validate(
        formData,
        schema,
        additionalMetaSchemas,
        customFormats
      );
      errors = schemaValidation.errors;
      errorSchema = schemaValidation.errorSchema;
      schemaValidationErrors = errors;
      schemaValidationErrorSchema = errorSchema;
    } else {
      const currentErrors = getCurrentErrors();
      errors = currentErrors.errors;
      errorSchema = currentErrors.errorSchema;
      schemaValidationErrors = state.schemaValidationErrors;
      schemaValidationErrorSchema = state.schemaValidationErrorSchema;
    }
    if (props.extraErrors) {
      errorSchema = mergeObjects(
        errorSchema,
        props.extraErrors,
        !!"concat arrays"
      );
      errors = toErrorList(errorSchema);
    }
    const idSchema = toIdSchema(
      retrievedSchema,
      uiSchema["ui:rootFieldId"],
      rootSchema,
      formData,
      props.idPrefix
    );
    const nextState = {
      schema,
      uiSchema,
      idSchema,
      formData,
      edit,
      errors,
      errorSchema,
      additionalMetaSchemas,
    };
    if (schemaValidationErrors) {
      nextState.schemaValidationErrors = schemaValidationErrors;
      nextState.schemaValidationErrorSchema = schemaValidationErrorSchema;
    }
    return nextState;
  }

  shouldComponentUpdate(nextProps, nextState) {
    return shouldRender(this, nextProps, nextState);
  }

  validate(
    formData,
    schema = this.props.schema,
    additionalMetaSchemas = this.props.additionalMetaSchemas,
    customFormats = this.props.customFormats
  ) {
    const { validate, transformErrors } = this.props;
    const { rootSchema } = this.getRegistry();
    const resolvedSchema = retrieveSchema(schema, rootSchema, formData);
    return validateFormData(
      formData,
      resolvedSchema,
      validate,
      transformErrors,
      additionalMetaSchemas,
      customFormats
    );
  }

  renderErrors() {
    const { errors, errorSchema, schema, uiSchema } = this.state;
    const { ErrorList, showErrorList, formContext } = this.props;

    if (errors.length && showErrorList != false) {
      return (
        <ErrorList
          errors={errors}
          errorSchema={errorSchema}
          schema={schema}
          uiSchema={uiSchema}
          formContext={formContext}
        />
      );
    }
    return null;
  }

  getUsedFormData = (formData, fields) => {
    //for the case of a single input form
    if (fields.length === 0 && typeof formData !== "object") {
      return formData;
    }

    let data = _pick(formData, fields);
    if (Array.isArray(formData)) {
      return Object.keys(data).map(key => data[key]);
    }

    return data;
  };

  getFieldNames = (pathSchema, formData) => {
    const getAllPaths = (_obj, acc = [], paths = [""]) => {
      Object.keys(_obj).forEach(key => {
        if (typeof _obj[key] === "object") {
          let newPaths = paths.map(path => `${path}.${key}`);
          // If an object is marked with additionalProperties, all its keys are valid
          if (_obj[key].__rjsf_additionalProperties && _obj[key].$name !== "") {
            acc.push(_obj[key].$name);
          } else {
            getAllPaths(_obj[key], acc, newPaths);
          }
        } else if (key === "$name" && _obj[key] !== "") {
          paths.forEach(path => {
            path = path.replace(/^\./, "");
            const formValue = _get(formData, path);
            // adds path to fieldNames if it points to a value
            // or an empty object/array
            if (typeof formValue !== "object" || _isEmpty(formValue)) {
              acc.push(path);
            }
          });
        }
      });
      return acc;
    };

    return getAllPaths(pathSchema);
  };

  onChange = (formData, newErrorSchema) => {
    if (isObject(formData) || Array.isArray(formData)) {
      const newState = this.getStateFromProps(this.props, formData);
      formData = newState.formData;
    }
    const mustValidate = !this.props.noValidate && this.props.liveValidate;
    let state = { formData };
    let newFormData = formData;

    if (this.props.omitExtraData === true && this.props.liveOmit === true) {
      const retrievedSchema = retrieveSchema(
        this.state.schema,
        this.state.schema,
        formData
      );
      const pathSchema = toPathSchema(
        retrievedSchema,
        "",
        this.state.schema,
        formData
      );

      const fieldNames = this.getFieldNames(pathSchema, formData);

      newFormData = this.getUsedFormData(formData, fieldNames);
      state = {
        formData: newFormData,
      };
    }

    if (mustValidate) {
      let schemaValidation = this.validate(newFormData);
      let errors = schemaValidation.errors;
      let errorSchema = schemaValidation.errorSchema;
      const schemaValidationErrors = errors;
      const schemaValidationErrorSchema = errorSchema;
      if (this.props.extraErrors) {
        errorSchema = mergeObjects(
          errorSchema,
          this.props.extraErrors,
          !!"concat arrays"
        );
        errors = toErrorList(errorSchema);
      }
      state = {
        formData: newFormData,
        errors,
        errorSchema,
        schemaValidationErrors,
        schemaValidationErrorSchema,
      };
    } else if (!this.props.noValidate && newErrorSchema) {
      const errorSchema = this.props.extraErrors
        ? mergeObjects(
            newErrorSchema,
            this.props.extraErrors,
            !!"concat arrays"
          )
        : newErrorSchema;
      state = {
        formData: newFormData,
        errorSchema: errorSchema,
        errors: toErrorList(errorSchema),
      };
    }
    this.setState(
      state,
      () => this.props.onChange && this.props.onChange(this.state)
    );
  };

  onBlur = (...args) => {
    if (this.props.onBlur) {
      this.props.onBlur(...args);
    }
  };

  onFocus = (...args) => {
    if (this.props.onFocus) {
      this.props.onFocus(...args);
    }
  };

  onSubmit = event => {
    event.preventDefault();
    if (event.target !== event.currentTarget) {
      return;
    }

    event.persist();
    let newFormData = this.state.formData;

    if (this.props.omitExtraData === true) {
      const retrievedSchema = retrieveSchema(
        this.state.schema,
        this.state.schema,
        newFormData
      );
      const pathSchema = toPathSchema(
        retrievedSchema,
        "",
        this.state.schema,
        newFormData
      );

      const fieldNames = this.getFieldNames(pathSchema, newFormData);

      newFormData = this.getUsedFormData(newFormData, fieldNames);
    }

    if (!this.props.noValidate) {
      let schemaValidation = this.validate(newFormData);
      let errors = schemaValidation.errors;
      let errorSchema = schemaValidation.errorSchema;
      const schemaValidationErrors = errors;
      const schemaValidationErrorSchema = errorSchema;
      if (Object.keys(errors).length > 0) {
        if (this.props.extraErrors) {
          errorSchema = mergeObjects(
            errorSchema,
            this.props.extraErrors,
            !!"concat arrays"
          );
          errors = toErrorList(errorSchema);
        }
        this.setState(
          {
            errors,
            errorSchema,
            schemaValidationErrors,
            schemaValidationErrorSchema,
          },
          () => {
            if (this.props.onError) {
              this.props.onError(errors);
            } else {
              console.error("Form validation failed", errors);
            }
          }
        );
        return;
      }
    }

    // There are no errors generated through schema validation.
    // Check for user provided errors and update state accordingly.
    let errorSchema;
    let errors;
    if (this.props.extraErrors) {
      errorSchema = this.props.extraErrors;
      errors = toErrorList(errorSchema);
    } else {
      errorSchema = {};
      errors = [];
    }

    this.setState(
      {
        formData: newFormData,
        errors: errors,
        errorSchema: errorSchema,
        schemaValidationErrors: [],
        schemaValidationErrorSchema: {},
      },
      () => {
        if (this.props.onSubmit) {
          this.props.onSubmit(
            { ...this.state, formData: newFormData, status: "submitted" },
            event
          );
        }
      }
    );
  };

  getRegistry() {
    // For BC, accept passed SchemaField and TitleField props and pass them to
    // the "fields" registry one.
    const { fields, widgets } = getDefaultRegistry();
    return {
      fields: { ...fields, ...this.props.fields },
      widgets: { ...widgets, ...this.props.widgets },
      ArrayFieldTemplate: this.props.ArrayFieldTemplate,
      ObjectFieldTemplate: this.props.ObjectFieldTemplate,
      FieldTemplate: this.props.FieldTemplate,
      definitions: this.props.schema.definitions || {},
      rootSchema: this.props.schema,
      formContext: this.props.formContext || {},
    };
  }

  submit() {
    if (this.formElement) {
      this.formElement.dispatchEvent(
        new CustomEvent("submit", {
          cancelable: true,
        })
      );
    }
  }

  render() {
    const {
      children,
      id,
      idPrefix,
      className,
      tagName,
      name,
      method,
      target,
      action,
      autocomplete: deprecatedAutocomplete,
      autoComplete: currentAutoComplete,
      enctype,
      acceptcharset,
      noHtml5Validate,
      disabled,
      readonly,
      formContext,
    } = this.props;

    const { schema, uiSchema, formData, errorSchema, idSchema } = this.state;
    const registry = this.getRegistry();
    const _SchemaField = registry.fields.SchemaField;
    const FormTag = tagName ? tagName : "form";
    if (deprecatedAutocomplete) {
      console.warn(
        "Using autocomplete property of Form is deprecated, use autoComplete instead."
      );
    }
    const autoComplete = currentAutoComplete
      ? currentAutoComplete
      : deprecatedAutocomplete;

    return (
      <FormTag
        className={className ? className : "rjsf"}
        id={id}
        name={name}
        method={method}
        target={target}
        action={action}
        autoComplete={autoComplete}
        encType={enctype}
        acceptCharset={acceptcharset}
        noValidate={noHtml5Validate}
        onSubmit={this.onSubmit}
        ref={form => {
          this.formElement = form;
        }}>
        {this.renderErrors()}
        <_SchemaField
          schema={schema}
          uiSchema={uiSchema}
          errorSchema={errorSchema}
          idSchema={idSchema}
          idPrefix={idPrefix}
          formContext={formContext}
          formData={formData}
          onChange={this.onChange}
          onBlur={this.onBlur}
          onFocus={this.onFocus}
          registry={registry}
          disabled={disabled}
          readonly={readonly}
        />
        {children ? (
          children
        ) : (
          <div>
            <button type="submit" className="btn btn-info">
              Submit
            </button>
          </div>
        )}
      </FormTag>
    );
  }
}

if (process.env.NODE_ENV !== "production") {
  Form.propTypes = {
    schema: PropTypes.object.isRequired,
    uiSchema: PropTypes.object,
    formData: PropTypes.any,
    disabled: PropTypes.bool,
    readonly: PropTypes.bool,
    widgets: PropTypes.objectOf(
      PropTypes.oneOfType([PropTypes.func, PropTypes.object])
    ),
    fields: PropTypes.objectOf(PropTypes.elementType),
    ArrayFieldTemplate: PropTypes.elementType,
    ObjectFieldTemplate: PropTypes.elementType,
    FieldTemplate: PropTypes.elementType,
    ErrorList: PropTypes.func,
    onChange: PropTypes.func,
    onError: PropTypes.func,
    showErrorList: PropTypes.bool,
    onSubmit: PropTypes.func,
    id: PropTypes.string,
    className: PropTypes.string,
    tagName: PropTypes.elementType,
    name: PropTypes.string,
    method: PropTypes.string,
    target: PropTypes.string,
    action: PropTypes.string,
    autocomplete: PropTypes.string,
    autoComplete: PropTypes.string,
    enctype: PropTypes.string,
    acceptcharset: PropTypes.string,
    noValidate: PropTypes.bool,
    noHtml5Validate: PropTypes.bool,
    liveValidate: PropTypes.bool,
    validate: PropTypes.func,
    transformErrors: PropTypes.func,
    formContext: PropTypes.object,
    customFormats: PropTypes.object,
    additionalMetaSchemas: PropTypes.arrayOf(PropTypes.object),
    omitExtraData: PropTypes.bool,
    extraErrors: PropTypes.object,
  };
}
