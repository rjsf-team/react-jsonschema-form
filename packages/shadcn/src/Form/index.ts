export { default } from './Form';
export * from './Form';
