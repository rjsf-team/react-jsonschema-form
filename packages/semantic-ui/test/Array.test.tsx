import arrayTests from '@rjsf/core/testSnap/arrayTests';

import Form from '../src';

/** Mock the `react-component-ref` component used by semantic-ui to simply render the children, otherwise tests fail */
jest.mock('@fluentui/react-component-ref', () => ({
  ...jest.requireActual('@fluentui/react-component-ref'),
  Ref: jest.fn().mockImplementation(({ children }) => children),
}));

arrayTests(Form);
