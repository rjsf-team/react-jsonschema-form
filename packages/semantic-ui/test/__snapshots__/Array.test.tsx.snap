// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`array fields array 1`] = `
<form
  className="ui form rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-array"
  >
    <div
      className="grouped equal width fields"
    >
      <div
        className="field field-array field-array-of-string sortable-form-fields"
      >
        <div>
          <div
            className="row array-item-list"
          />
          <div
            style={
              Object {
                "marginTop": "1rem",
                "position": "relative",
                "textAlign": "right",
              }
            }
          >
            <button
              className="ui tiny icon button"
              onClick={[Function]}
              title="Add Item"
            >
              <i
                aria-hidden="true"
                className="plus icon"
                onClick={[Function]}
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    className="ui primary button"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`array fields array icons 1`] = `
<form
  className="ui form rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-array"
  >
    <div
      className="grouped equal width fields"
    >
      <div
        className="field field-array field-array-of-string sortable-form-fields"
      >
        <div>
          <div
            className="row array-item-list"
          >
            <div
              className="array-item"
            >
              <div
                className="ui grid"
                style={
                  Object {
                    "display": "grid",
                    "gridTemplateColumns": "1fr 65px",
                  }
                }
              >
                <div
                  className="middle aligned sixteen wide column"
                >
                  <div
                    className="form-group field field-string"
                  >
                    <div
                      className="grouped equal width fields"
                    >
                      <div
                        className="required field"
                      >
                        <div
                          className="ui fluid input"
                        >
                          <input
                            aria-describedby="root_0__error root_0__description root_0__help"
                            autoFocus={false}
                            disabled={false}
                            id="root_0"
                            name="root_0"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            placeholder=""
                            required={true}
                            type="text"
                            value="a"
                          />
                        </div>
                      </div>
                      
                    </div>
                  </div>
                </div>
                <div
                  className="column"
                >
                  <div
                    className="ui mini vertical buttons"
                  >
                    <button
                      className="ui icon disabled button array-item-move-up"
                      disabled={true}
                      onClick={[Function]}
                      tabIndex={-1}
                      title="Move up"
                    >
                      <i
                        aria-hidden="true"
                        className="angle up icon"
                        onClick={[Function]}
                      />
                    </button>
                    <button
                      className="ui icon button array-item-move-down"
                      onClick={[Function]}
                      title="Move down"
                    >
                      <i
                        aria-hidden="true"
                        className="angle down icon"
                        onClick={[Function]}
                      />
                    </button>
                    <button
                      className="ui icon button array-item-remove"
                      onClick={[Function]}
                      title="Remove"
                    >
                      <i
                        aria-hidden="true"
                        className="trash icon"
                        onClick={[Function]}
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="array-item"
            >
              <div
                className="ui grid"
                style={
                  Object {
                    "alignItems": "center",
                    "display": "grid",
                    "gridTemplateColumns": "1fr 65px",
                  }
                }
              >
                <div
                  className="middle aligned sixteen wide column"
                >
                  <div
                    className="form-group field field-string"
                  >
                    <div
                      className="grouped equal width fields"
                    >
                      <div
                        className="required field"
                      >
                        <div
                          className="ui fluid input"
                        >
                          <input
                            aria-describedby="root_1__error root_1__description root_1__help"
                            autoFocus={false}
                            disabled={false}
                            id="root_1"
                            name="root_1"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            placeholder=""
                            required={true}
                            type="text"
                            value="b"
                          />
                        </div>
                      </div>
                      
                    </div>
                  </div>
                </div>
                <div
                  className="column"
                >
                  <div
                    className="ui mini vertical buttons"
                  >
                    <button
                      className="ui icon button array-item-move-up"
                      onClick={[Function]}
                      title="Move up"
                    >
                      <i
                        aria-hidden="true"
                        className="angle up icon"
                        onClick={[Function]}
                      />
                    </button>
                    <button
                      className="ui icon disabled button array-item-move-down"
                      disabled={true}
                      onClick={[Function]}
                      tabIndex={-1}
                      title="Move down"
                    >
                      <i
                        aria-hidden="true"
                        className="angle down icon"
                        onClick={[Function]}
                      />
                    </button>
                    <button
                      className="ui icon button array-item-remove"
                      onClick={[Function]}
                      title="Remove"
                    >
                      <i
                        aria-hidden="true"
                        className="trash icon"
                        onClick={[Function]}
                      />
                    </button>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <div
            style={
              Object {
                "marginTop": "1rem",
                "position": "relative",
                "textAlign": "right",
              }
            }
          >
            <button
              className="ui tiny icon button"
              onClick={[Function]}
              title="Add Item"
            >
              <i
                aria-hidden="true"
                className="plus icon"
                onClick={[Function]}
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    className="ui primary button"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`array fields checkboxes 1`] = `
<form
  className="ui form rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-array"
  >
    <div
      className="grouped equal width fields"
    >
      <div
        className="field"
      >
        <div
          aria-describedby="root__error root__description root__help"
          aria-disabled={false}
          aria-expanded={false}
          aria-multiselectable={true}
          autoFocus={false}
          className="ui fluid multiple selection scrolling dropdown"
          id="root"
          inverted="false"
          name="root"
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onMouseDown={[Function]}
          readOnly={false}
          required={false}
          role="listbox"
          tabIndex={0}
        >
          <i
            aria-hidden="true"
            className="dropdown icon"
            onClick={[Function]}
          />
          <div
            className="menu transition"
          >
            <div
              aria-checked={false}
              aria-disabled={false}
              aria-selected={true}
              className="selected item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                a
              </span>
            </div>
            <div
              aria-checked={false}
              aria-disabled={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                b
              </span>
            </div>
            <div
              aria-checked={false}
              aria-disabled={false}
              aria-selected={false}
              className="item"
              onClick={[Function]}
              role="option"
              style={
                Object {
                  "pointerEvents": "all",
                }
              }
            >
              <span
                className="text"
              >
                c
              </span>
            </div>
          </div>
        </div>
      </div>
      
    </div>
  </div>
  <button
    className="ui primary button"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`array fields empty errors array 1`] = `
<form
  className="ui form rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <div
      className="grouped equal width fields"
    >
      
      <div
        className="form-group field field-string"
      >
        <div
          className="grouped equal width fields"
        >
          <div
            className="field"
          >
            <label
              htmlFor="root_name"
            >
              name
            </label>
            <div
              className="ui fluid input"
            >
              <input
                aria-describedby="root_name__error root_name__description root_name__help"
                autoFocus={false}
                disabled={false}
                id="root_name"
                name="root_name"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="text"
                value=""
              />
            </div>
          </div>
          
        </div>
      </div>
    </div>
  </div>
  <button
    className="ui primary button"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`array fields fixed array 1`] = `
<form
  className="ui form rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-array"
  >
    <div
      className="grouped equal width fields"
    >
      <div
        className="field field-array field-array-fixed-items"
      >
        <div>
          <div
            className="row array-item-list"
          >
            <div
              className="array-item"
            >
              <div
                className="ui grid"
                style={
                  Object {
                    "display": "grid",
                    "gridTemplateColumns": "1fr 65px",
                  }
                }
              >
                <div
                  className="middle aligned sixteen wide column"
                >
                  <div
                    className="form-group field field-string"
                  >
                    <div
                      className="grouped equal width fields"
                    >
                      <div
                        className="required field"
                      >
                        <div
                          className="ui fluid input"
                        >
                          <input
                            aria-describedby="root_0__error root_0__description root_0__help"
                            autoFocus={false}
                            disabled={false}
                            id="root_0"
                            name="root_0"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            placeholder=""
                            required={true}
                            type="text"
                            value=""
                          />
                        </div>
                      </div>
                      
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              className="array-item"
            >
              <div
                className="ui grid"
                style={
                  Object {
                    "alignItems": "center",
                    "display": "grid",
                    "gridTemplateColumns": "1fr 65px",
                  }
                }
              >
                <div
                  className="middle aligned sixteen wide column"
                >
                  <div
                    className="form-group field field-number"
                  >
                    <div
                      className="grouped equal width fields"
                    >
                      <div
                        className="required field"
                      >
                        <div
                          className="ui fluid input"
                        >
                          <input
                            aria-describedby="root_1__error root_1__description root_1__help"
                            autoFocus={false}
                            disabled={false}
                            id="root_1"
                            name="root_1"
                            onBlur={[Function]}
                            onChange={[Function]}
                            onFocus={[Function]}
                            placeholder=""
                            required={true}
                            step="any"
                            type="number"
                            value=""
                          />
                        </div>
                      </div>
                      
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <button
    className="ui primary button"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</form>
`;

exports[`array fields no errors 1`] = `
<form
  className="ui form rjsf"
  noValidate={false}
  onSubmit={[Function]}
>
  <div
    className="form-group field field-object"
  >
    <div
      className="grouped equal width fields"
    >
      
      <div
        className="form-group field field-string"
      >
        <div
          className="grouped equal width fields"
        >
          <div
            className="field"
          >
            <label
              htmlFor="root_name"
            >
              name
            </label>
            <div
              className="ui fluid input"
            >
              <input
                aria-describedby="root_name__error root_name__description root_name__help"
                autoFocus={false}
                disabled={false}
                id="root_name"
                name="root_name"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="text"
                value=""
              />
            </div>
          </div>
          
        </div>
      </div>
    </div>
  </div>
  <button
    className="ui primary button"
    onClick={[Function]}
    type="submit"
  >
    Submit
  </button>
</form>
`;
